seq2seq/metrics/metrics.py

```diff
@@ -356,4 +356,297 @@ def eval_batch(self, outputs, targets):
+
+        super(PonderTokenMetric, self).__init__(self._NAME, self._SHORTNAME, self._INPUT)
+
+    def get_val(self):
+        """
+        Get the average accuracy metric of all processed batches
+        Returns:
+            float: average accuracy
+        """
+        if self.ponder_total != 0:
+            return float(self.ponder_correct) / self.ponder_total
+        else:
+            return 0
+
+    def reset(self):
+        """
+        Reset after all batches have been processed
+        """
+        self.ponder_correct = 0
+        self.ponder_total = 0
+
+
+    def eval_batch(self, outputs, targets):
+        """
+        Evaluates one batch of inputs (grammar) and checks whether the predictions are correct in the
+        specified grammar.
+        Note that we assume that the input grammar's do not contain any EOS-like symbol
+        Args:
+            outputs (list(torch.tensor)): Contains the predictions of the model. List of length max_output_length, where each element is a tensor of length batch_size
+            targets (dict): Dictionary containing the grammars
+        """
+
+        # batch_size X N variable containing the indices of the model's input,
+        # where N is the longest input
+        input_variable = targets['encoder_input']
+        output_variable = targets['decoder_output']
+        batch_size = input_variable.size(0)
+
+        # Convert to batch_size x M variable containing the indices of the model's output, where M
+        # is the longest output
+        predictions = torch.stack(outputs, dim=1).view(batch_size, -1)
+
+        # Current implementation does not allow batch-wise evaluation
+        for i_batch_element in range(batch_size):
+            # We go through multiple checks for incorrectness.
+            # If all these test fail, we consider the sequence correct.
+
+            # Extract the current example and move to cpu
+            grammar = input_variable[i_batch_element, :].data.cpu().numpy()
+            prediction = predictions[i_batch_element, :].data.cpu().numpy()
+            target_output = output_variable[i_batch_element, :].data.cpu().numpy()
+
+
+            # Convert indices to strings
+            # Remove all padding from the grammar.
+            grammar = [self.input_vocab.itos[token] for token in grammar if
+                       self.input_vocab.itos[token] != self.input_pad_symbol]
+            prediction = [self.output_vocab.itos[token] for token in prediction]
+            target_output = [self.output_vocab.itos[token] for token in target_output if
+                          self.output_vocab.itos[token] != self.output_pad_symbol ]
+
+
+            # # The first prediction after the actual output should be EOS
+            if self.use_output_eos and prediction[-1] != self.output_eos_symbol:
+                continue
+
+            # Remove EOS (and possible padding)
+            ponder_length = prediction[:prediction.index(self.output_eos_symbol)]
```

Comment:
This assumes `self.use_output_eos` to be `True`

Reply:
changed this to:

ponder_length = prediction[:prediction.index(self.output_eos_symbol)] if self.use_output_eos else prediction
tgt_length = target_output[1:-1] if self.use_output_eos else target_output